search_agg_percentiles.go

```diff
@@ -49,21 +50,109 @@ func (p *PercentilesAggregation) ToOpenSearchJSON() ([]byte, error) {
 	return json.Marshal(source)
 }
 
-// PercentileAggregationResult will contain all percentiles or no percentiles.
+// PercentilesAggregationResult will contain all percentiles or no percentiles.
 // If there are no values for the percentile, it will be omitted
-type PercentileAggregationResult struct {
-	P1        *float64 `json:"1.0,omitempty"`
-	P1String  string   `json:"1.0_as_string,omitempty"`
-	P5        *float64 `json:"5.0,omitempty"`
-	P5String  string   `json:"5.0_as_string,omitempty"`
-	P25       *float64 `json:"25.0,omitempty"`
-	P25String string   `json:"25.0_as_string,omitempty"`
-	P50       *float64 `json:"50.0,omitempty"`
-	P50String string   `json:"50.0_as_string,omitempty"`
-	P75       *float64 `json:"75.0,omitempty"`
-	P75String string   `json:"75.0_as_string,omitempty"`
-	P95       *float64 `json:"95.0,omitempty"`
-	P95String string   `json:"95.0_as_string,omitempty"`
-	P99       *float64 `json:"99.0,omitempty"`
-	P99String string   `json:"99.0_as_string,omitempty"`
+type PercentilesAggregationResult struct {
+	P1        *float64
+	P1String  string
+	P5        *float64
+	P5String  string
+	P25       *float64
+	P25String string
+	P50       *float64
+	P50String string
+	P75       *float64
+	P75String string
+	P95       *float64
+	P95String string
+	P99       *float64
+	P99String string
+}
+
+// UnmarshalJSON implements [json.Unmarshaler] to decode a json byte slice into a PercentilesAggregationResult
+func (p *PercentilesAggregationResult) UnmarshalJSON(b []byte) error {
+	// map[key] -> value
+	var rawResp map[string]json.RawMessage
+	if err := json.Unmarshal(b, &rawResp); err != nil {
+		return err
+	}
+
+	if p == nil {
+		return fmt.Errorf("invalid PercentilesAggregationResult target, nil")
+	}
+
+	valueMessage, ok := rawResp["values"]
```

Comment:
So the result structure is a top-level `values` key with nested fields for each percentile?
```
{"values": {"1.0": 0.01, "1.0_as_string": "0.01", ...}}
```

If it's a stable structure you can use an "auxiliary" struct with JSON tags and still unmarshal things directly.
```
type parJSON struct {
    Values struct {
        P1 *float64 `json:"1.0,omitempty"`
        P1String string `json:"1.0_as_string,omitempty"`
        ...
    } `json:"values"`
}
var aux parJSON
err := json.Unmarshal(data, &aux)
...
// copy the values from aux to p
```

I feel like we should limit manually looping over JSON keys to the cases where the set of keys is variable/dynamic.

Reply:
Yea. I started with auxiliary structures at first, but then moved into this loop pattern.

For this one in specific, those are the only percentiles that will be returned. If you'd prefer the aux approach I'm happy to make that change though. Thoughts?

Reply:
Mostly thinking that this is an uncommon pattern so we should only use it where we **must**.  If we can define a struct that lines up with the response then we should do that.

Reply:
changed